python-checks/src/main/java/org/sonar/python/checks/TfFunctionRecursivityCheck.java

```diff
@@ -0,0 +1,114 @@
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 3 of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+package org.sonar.python.checks;
+
+import java.util.ArrayList;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Objects;
+import java.util.Optional;
+import java.util.Set;
+import javax.annotation.Nonnull;
+import org.sonar.check.Rule;
+import org.sonar.plugins.python.api.PythonSubscriptionCheck;
+import org.sonar.plugins.python.api.SubscriptionContext;
+import org.sonar.plugins.python.api.symbols.Symbol;
+import org.sonar.plugins.python.api.symbols.Usage;
+import org.sonar.plugins.python.api.tree.BaseTreeVisitor;
+import org.sonar.plugins.python.api.tree.CallExpression;
+import org.sonar.plugins.python.api.tree.Decorator;
+import org.sonar.plugins.python.api.tree.Expression;
+import org.sonar.plugins.python.api.tree.FunctionDef;
+import org.sonar.plugins.python.api.tree.Tree;
+import org.sonar.python.tree.TreeUtils;
+
+@Rule(key = "S6908")
+public class TfFunctionRecursivityCheck extends PythonSubscriptionCheck {
+
+  private static final String MESSAGE = "Remove this recursive call.";
+  private static final String SECONDARY_MESSAGE = "Recursive call is here.";
+
+  @Override
+  public void initialize(Context context) {
+    context.registerSyntaxNodeConsumer(Tree.Kind.FUNCDEF, TfFunctionRecursivityCheck::checkFunctionDef);
+  }
+
+  private static void checkFunctionDef(SubscriptionContext context) {
+    FunctionDef functionDef = (FunctionDef) context.syntaxNode();
+    if (!isTfFunction(functionDef)) {
+      return;
+    }
+    var selfSymbol = TreeUtils.getFunctionSymbolFromDef(functionDef);
+    if (selfSymbol == null) {
+      return;
+    }
+    var collector = new CallCollector(selfSymbol);
+    context.syntaxNode().accept(collector);
+    if (collector.expressionList.isEmpty()) {
+      return;
+    }
+    var issue = context.addIssue(functionDef.name(), MESSAGE);
+    collector.expressionList.forEach(call -> issue.secondary(call, SECONDARY_MESSAGE));
+  }
+
+  private static boolean isTfFunction(FunctionDef functionDefinition) {
+    return functionDefinition.decorators().stream()
+      .map(Decorator::expression).map(TreeUtils::getSymbolFromTree)
+      .filter(Optional::isPresent).map(Optional::get)
+      .map(Symbol::fullyQualifiedName)
+      .filter(Objects::nonNull)
+      .anyMatch("tensorflow.function"::equals);
```

Comment:
I must say I don't really like this construct, as this kind of construct has caused some NPEs in production in the past.

In particular, I believe the `Objects::nonNull` filter is not actually covered in the tests (no `null` object ever gets through it, although I believe it can be triggered). If you forgot the filter and performed an unsafe operation afterward, you could get a NPE that may go unseen until we reach production. I believe this code is actually safe, but it is quite implicit and untested.

I believe extracting the `isTfFunctionDecorator` bit in a dedicated method (as well as adding the relevant tests for other kind of decorators) may help achieving safer code. It would also avoid the awkward `Optional::isPresent / Optional::get` part. Something that would look like:

```
private static boolean isTfFunctionDecorator(Decorator decorator) {
    return Optional.of(decorator.expression())
      .flatMap(TreeUtils::getSymbolFromTree)
      .map(Symbol::fullyQualifiedName)
      .filter("tensorflow.function"::equals)
      .isPresent();
  }
```

(Note that even with this implementation, we have blind spots as to which of these steps can actually return `null` and we shouldn't just rely on the 100% coverage to leave these cases uncovered)